examples/neighbors/approximate_nearest_neighbors.py

```diff
@@ -0,0 +1,289 @@
+"""
+=====================================
+Approximate nearest neighbors in TSNE
+=====================================
+
+This example presents how to chain KNeighborsTransformer and TSNE in a
+pipeline, and how to wrap the packages `annoy` and `nmslib` to replace
+KNeighborsTransformer and perform approximate nearest neighbors.
+These package can be installed with `pip install annoy nmslib`.
+
+Note: Currently TSNE(metric='precomputed') does not modify the precomputed
+distances, and thus assumes that precomputed euclidean distances are squared.
+In future versions, a parameter in TSNE will control the optional squaring of
+precomputed distances (see #12401).
+
+Sample output:
+
+Benchmarking on MNIST_2000:
+---------------------------
+AnnoyTransformer:                    0.583 sec
+NMSlibTransformer:                   0.321 sec
+KNeighborsTransformer:               1.225 sec
+TSNE with AnnoyTransformer:          4.903 sec
+TSNE with NMSlibTransformer:         5.009 sec
+TSNE with KNeighborsTransformer:     6.210 sec
```

Comment:
why is there a difference between the last two?

Reply:
Different overheads ?
On three runs:
```
TSNE with KNeighborsTransformer:     6.170 sec
TSNE with internal NearestNeighbors: 6.159 sec
...
TSNE with KNeighborsTransformer:     50.391 sec
TSNE with internal NearestNeighbors: 50.596 sec
```
```
TSNE with KNeighborsTransformer:     6.187 sec
TSNE with internal NearestNeighbors: 6.201 sec
...
TSNE with KNeighborsTransformer:     50.655 sec
TSNE with internal NearestNeighbors: 50.832 sec
```
```
TSNE with KNeighborsTransformer:     6.155 sec
TSNE with internal NearestNeighbors: 6.173 sec
...
TSNE with KNeighborsTransformer:     50.574 sec
TSNE with internal NearestNeighbors: 51.156 sec
```